### PR TITLE
Added document detailing org roadmap to Core 1.0.0

### LINKED ISSUE #157: Create document detailing pre-reqs for an InstructLab CLI 1.0.0
We want to make a public statement as to what features/designs/integrations/etc we, as the Maintainer team, want to see in the project prior to a 1.0.0 release

### PR DESCRIPTION
**Pull request by nathan-weinberg**
Resolves #157

### CODE REVIEW

**Comment by nathan-weinberg** [comment]
@RobotSail I didn't include the UI piece in this doc since it's only for the CLI, but let me know if you wanna talk that out or have me mention something in the doc somewhere about a UI anyway, maybe under the Q&A

**Comment by RobotSail** [comment]
@nathan-weinberg Ah gotcha, we can find a better place for it.

**Review by cdoern — COMMENTED** [review]
just two small comments, I like the very general approach here.

File: docs/instructlab-cli-1.0.0.md
Comment (JamesKunstle) [inline docs/instructlab-cli-1.0.0.md:21]: IMO this is a really big commitment that we don't necessarily have to make. It's common for consumers of python projects to expect minimal backward compat and to take the most recent version as that one with fewest feature errors.

File: docs/instructlab-cli-1.0.0.md
Comment (JamesKunstle) [inline docs/instructlab-cli-1.0.0.md:22]: Good idea

File: docs/instructlab-cli-1.0.0.md
Comment (JamesKunstle) [inline docs/instructlab-cli-1.0.0.md:23]: I would also add sufficient unit and integration testing from the subcomponents. That's something I care about building for `instructlab/training`. A user of a "1.0" project should be able to run most of the tests quickly on their machine and cover most behavior.

File: docs/instructlab-cli-1.0.0.md
Comment (JamesKunstle) [reply docs/instructlab-cli-1.0.0.md:28]: We'll need to plan this but I agree, the libraries will have to catch up to the backgroundable/RESTy version of ilab.

File: docs/instructlab-cli-1.0.0.md
Comment (JamesKunstle) [inline docs/instructlab-cli-1.0.0.md:1]: I'm not anxious to push for a 1.0.0 release for the sake of **having** a >1.0.0 project. I don't personally believe that semver major versions are all that meaningful to the end user. 

However, these are good objectives for us and would represent a major change in the implementation and design of the project, which I'm certainly on board with.

File: docs/instructlab-cli-1.0.0.md
Comment (bbrowning) [reply docs/instructlab-cli-1.0.0.md:28]: I don't think the CLI would be able to meet any of its backwards compatibility goals stated here unless the libraries also committed to those goals? This is perhaps a more general issue outside of the scope of just a CLI 1.0.0 as to what level of maturity and backwards compatibility of our Python APIs we want to commit to across repos in the org, as that ultimately determines what level of backwards compatibility consumers of those APIs (like CLI) can commit to.

File: docs/instructlab-cli-1.0.0.md
Comment (jwm4) [inline docs/instructlab-cli-1.0.0.md:7]: One thing that is missing from this context section that came up in the discussion of the SDG refactoring draft dev doc is the fact that the repo used to be `instructlab/cli` and largely focused on just the CLI layer and now it is `instructlab/instructlab` and is more of a core integrating element that contains a lot of the glue functionality for all of InstructLab.  I think this document should either state this or link to a document that states this.

I also have a proposal that I think might be more controversial.  Can we stop calling this component the CLI and provide a new name for it?   The repo could still be `instructlab/instructlab` but we would need to update the README, the upstream Slack channel name, start using the new name in dev docs and Jira issues, update the labels in meeting invites, etc.; so it is not a small undertaking, but I think it is one that will provide a lot of value in the long term because having accurate names for the major components makes it easier for new participants in the project to understand the system.

I think InstructLab core would be a good name, but I would be open to other names that convey a sense that the is the central orchestrator for all of the components of InstructLab .

Finally, I really think the name issue should be sorted out in this dev doc because settling on a name and providing consistent communication around the name should be a prereq for declaring the repo 1.0.0.

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:28]: also a good point @bbrowning - what do you think might be a reasonable expectation, both for ourselves and the community?

File: docs/instructlab-cli-1.0.0.md
Comment (jwm4) [reply docs/instructlab-cli-1.0.0.md:21]: Would a compromise be to commit to backward compatibility across all 1.0.y versions?  I.e., that API breaking changes will not happen on third digit point releases.

Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:21]: Good point, the vagueness of this makes it quite the lift - maybe we can commit to backwards compatibility for N-1 Y-Streams?

E.g., `1.4` is backwards-compatible with `1.3` but not _necessarily_ `1.2` 

Or do you think even that might be more than worth committing to?

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:22]: @kelbrown20 is actually working on this now! https://github.com/instructlab/instructlab/pull/2670

**Review by jwm4 — CHANGES_REQUESTED** [review]
There seem to be a few open issues here especially around naming and backward compatibility.

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:23]: I very much agree with the sentiment - can you think of any concrete milestones we may want to define here? I do want to keep this confined mainly to the CLI, so keeping it moreso in that context than the libraries

File: docs/instructlab-cli-1.0.0.md
Comment (jwm4) [reply docs/instructlab-cli-1.0.0.md:21]: N-1-Y sounds complicated to manage to me because breaking changes need to be staged out across multiple releases.

File: docs/instructlab-cli-1.0.0.md
Comment (junaruga) [inline docs/instructlab-cli-1.0.0.md:18]: This REST API looks nice. I think this encourages people's contributions, and people with a less power laptop can train the dataset faster.

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:21]: Def good with backwards compatibility across Z-Streams only - @JamesKunstle wdyt?

**Comment by anastasds** [comment]
Thanks for putting this together!

I suggest considering architectural goals even if they are implied by things already listed, e.g. the REST API work will probably drive modularization. Do we to stabilize our systems architecture before declaring a 1.0?

What about API guarantees? State of CI? Documentation? Testing?

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:7]: > One thing that is missing from this context section that came up in the discussion of the SDG refactoring draft dev doc is the fact that the repo used to be instructlab/cli and largely focused on just the CLI layer and now it is instructlab/instructlab and is more of a core integrating element that contains a lot of the glue functionality for all of InstructLab. I think this document should either state this or link to a document that states this.

I agree with this sentiment, though candidly I don't know where the decision of record is for this - I'll need to do some digging.

> I also have a proposal that I think might be more controversial. Can we stop calling this component the CLI and provide a new name for it? The repo could still be instructlab/instructlab but we would need to update the README, the upstream Slack channel name, start using the new name in dev docs and Jira issues, update the labels in meeting invites, etc.; so it is not a small undertaking, but I think it is one that will provide a lot of value in the long term because having accurate names for the major components makes it easier for new participants in the project to understand the system.

I can add a section about this and update the existing references - I agree that the "CLI" is really just a part of the overall `instructlab` package that serves as a means for controlling the orchestration, which will in the future have other entrypoints such as a GUI and direct API calls. That being said, I don't feel an entirely holistic renaming of things here is necessary - some things like parts of the README and the meeting name yes, but the CLI portion of this component is always going to exist and need a place for discussion/way to be referred to.

> I think InstructLab core would be a good name, but I would be open to other names that convey a sense that the is the central orchestrator for all of the components of InstructLab .

Engine might also be a good term for it

> Finally, I really think the name issue should be sorted out in this dev doc because settling on a name and providing consistent communication around the name should be a prereq for declaring the repo 1.0.0.

Let's definitely flesh this out in the document - I'll add a new section for it. That being said, I am wary this issue might bog down the rest of the document, and reserve the right to postpone a final decision to after merging this, while specifying that a decision _must_ be made prior to 1.0.0.

**Comment by nathan-weinberg** [comment]
> Thanks for putting this together!
> 
> I suggest considering architectural goals even if they are implied by things already listed, e.g. the REST API work will probably drive modularization. Do we want to stabilize our systems architecture before declaring a 1.0?
> 
> What about API guarantees? State of CI? Documentation? Testing? I think it can definitely be reasonable for these things to be out of scope for this doc - but if so, I think it would be prudent to say so and at least hint at how and where those will end up being addressed.

Of course! I definitely agree with the sentiment - I have some concrete items under the `MVP for an InstructLab CLI 1.0.0` - I will make some additions there and we can further refine them

**Comment by anastasds** [comment]
I'm happy to help discuss architectural goals if we want to scope those into this document.

I would also point out that what you've put together here is in the direction of defining a technical vision to guide us for somewhere between 6 and 12 months probably. More would need to be done to make it that, but this a first step on the way there. I think it would be appropriate to spend some time on this the next time the org gets together in a planning meeting or some such - it might deserve its own one-off meeting.

Next steps - which would not be in the scope of this document - would be to work backwards from whatever goals are defined here to identify milestones etc with buy-in from the whole org.

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [inline docs/instructlab-cli-1.0.0.md:3]: 😆

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [inline docs/instructlab-cli-1.0.0.md:13]: "behoove"
"to wit"

A++++ would buy again

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [reply docs/instructlab-cli-1.0.0.md:17]: I think including both makes sense!

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [inline docs/instructlab-cli-1.0.0.md:18]: Was this in doubt? What is being suggested here, a governance change?

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [inline docs/instructlab-cli-1.0.0.md:26]: Some projects name this *-core. Eg was looking up jupyter packages in fedora today... they have a jupyter-core package. Eg. https://packages.fedoraproject.org/search?query=*core and https://pypi.org/search/?q=*-core

*-base also appears popular

https://packages.fedoraproject.org/search?query=*base and https://pypi.org/search/?q=*-base (first page skip on pypi)

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [inline docs/instructlab-cli-1.0.0.md:28]: 👍

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [inline docs/instructlab-cli-1.0.0.md:42]: We should probably better define what this actually means.

File: docs/instructlab-cli-1.0.0.md
Comment (mairin) [reply docs/instructlab-cli-1.0.0.md:58]: Can we not say we have achieved a certain phase of the CI improvement plan before we can declare 1.0.0? (if one of the goals is stability....)

File: docs/instructlab-cli-1.0.0.md
Comment (RobotSail) [reply docs/instructlab-cli-1.0.0.md:42]: +1

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:18]: Definitely not in doubt, I just think it more signifies our existing commitment - a 1.0+ piece of software is less likely to be abandoned and more likely to be maintained than a pre-1.0 piece of software (maybe that's my own personal bias :smile:)

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:26]: I believe @jwm4 had suggested Core as well!

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:42]: I would love for the folks working on RAG to chime in here and suggest more detail, I've been pretty far removed from that work

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:58]: Hmmm yes I like that, I can add a point around this!

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:17]: Will update with this language!

**Comment by nathan-weinberg** [comment]
Possible new names for "CLI" (note "CLI" will continued to be used for specific CLI discussions, just not the `instructlab/instructlab` repo at large) 

- Core (react :heart:)
- Engine (react :+1:)
- Orchestrator (react :tada:)

Feel free to respond with a different suggestion as well!

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md:42]: I am going to resolve this convo for now since no SMEs have chimed in - we can flesh this out in a followup

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md]: Barring any input, this will go in as-is on Friday - can update it in a followup

File: docs/instructlab-cli-1.0.0.md
Comment (nathan-weinberg) [reply docs/instructlab-cli-1.0.0.md]: Barring any input, this will go in as-is on Friday - can update it in a followup

**Comment by nathan-weinberg** [comment]
> Possible new names for "CLI" (note "CLI" will continued to be used for specific CLI discussions, just not the `instructlab/instructlab` repo at large)
> 
>     * Core (react ❤️)
> 
>     * Engine (react 👍)
> 
>     * Orchestrator (react 🎉)
> 
> 
> Feel free to respond with a different suggestion as well!

At this time, voting has closed with `Core` being the winner

**Comment by nathan-weinberg** [comment]
As previously communicated, I am going to merge this

Any desired changes may be proposed as follow-ups

**Pull request merged**
